File: components/input/android/java/src/org/chromium/components/input/InputFeatureMap.java
// Copyright 2025 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.input;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.base.FeatureMap;
import org.chromium.build.annotations.NullMarked;

/** Java accessor for state of Input feature flags. */
@JNINamespace("input")
@NullMarked
public class InputFeatureMap extends FeatureMap {
    public static final String USE_ANDROID_BUFFERED_INPUT_DISPATCH =
            "UseAndroidBufferedInputDispatch";

    private static final InputFeatureMap sInstance = new InputFeatureMap();

    // Do not instantiate this class.
    private InputFeatureMap() {
        super();
    }

    /**
     * @return the singleton {@link InputFeatureMap}.
     */
    public static InputFeatureMap getInstance() {
        return sInstance;
    }

    /** Convenience method to call {@link #isEnabledInNative(String)} statically. */
    public static boolean isEnabled(String featureName) {
        return getInstance().isEnabledInNative(featureName);
    }

    @Override
    protected long getNativeMap() {
        return InputFeatureMapJni.get().getNativeMap();
    }

    @NativeMethods
    interface Natives {
        long getNativeMap();
    }
}

File: components/input/android/input_feature_map.cc
// Copyright 2025 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "base/android/feature_map.h"
#include "base/feature_list.h"
#include "base/no_destructor.h"
#include "components/input/features.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "components/input/android/jni_headers/InputFeatureMap_jni.h"

namespace input {

namespace {

// Array of features exposed through the Java InputFeatureMap API.
const base::Feature* const kFeaturesExposedToJava[] = {
    &input::features::kUseAndroidBufferedInputDispatch,
};

// static
base::android::FeatureMap* GetFeatureMap() {
  static base::NoDestructor<base::android::FeatureMap> kFeatureMap(
      kFeaturesExposedToJava);
  return kFeatureMap.get();
}

}  // namespace

static jlong JNI_InputFeatureMap_GetNativeMap(JNIEnv* env) {
  return reinterpret_cast<jlong>(GetFeatureMap());
}

}  // namespace input

DEFINE_JNI(InputFeatureMap)
